src/librustc_mir/transform/const_prop.rs

```diff
@@ -813,7 +826,7 @@ impl<'tcx> Visitor<'tcx> for CanConstProp {
             | MutatingUse(MutatingUseContext::Borrow)
             | MutatingUse(MutatingUseContext::AddressOf) => {
                 trace!("local {:?} can't be propagaged because it's used: {:?}", local, context);
-                self.can_const_prop[local] = ConstPropMode::OnlyPropagateInto;
+                self.can_const_prop[local] = ConstPropMode::NoPropagation;
```

Comment:
Main part of the bugfix, while we can do things for borrows, this is super hard to reason about and I don't think we should do it without dataflow